### PR TITLE
C#: CSV-based flow summaries

### PR DESCRIPTION
**Pull request by hvitved**
~https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1138/~
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1145/

### CODE REVIEW

**Review by tamasvajk — APPROVED** [review]
Looks good to me.

I'll add the [Coverage.ql](https://github.com/github/codeql/blob/main/java/ql/src/meta/frameworks/Coverage.ql), and then modify the CSV coverage report workflow to handle C# too.

File: csharp/ql/test/library-tests/dataflow/external-models/steps.ql [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        "My.Qltest;C;false;StepQualRes;();;Argument[-1];ReturnValue;taint",
+        "My.Qltest;C;false;StepQualArg;(object);;Argument[-1];Argument[0];taint",
+        "My.Qltest;C;false;StepFieldGetter;();;Field[My.Qltest.C.Field] of Argument[-1];ReturnValue;value",
+        "My.Qltest;C;false;StepFieldSetter;(int);;Argument[0];Field[My.Qltest.C.Field] of Argument[-1];value",
+        "My.Qltest;C;false;StepPropertyGetter;();;Property[My.Qltest.C.Property] of Argument[-1];ReturnValue;value",
+        "My.Qltest;C;false;StepPropertySetter;(int);;Argument[0];Property[My.Qltest.C.Property] of Argument[-1];value",
+        "My.Qltest;C;false;StepElementGetter;();;Element of Argument[-1];ReturnValue;value",
+        "My.Qltest;C;false;StepElementSetter;(int);;Argument[0];Element of Argument[-1];value",
+        "My.Qltest.C;Generic<>;false;StepGeneric;(T);;Argument[0];ReturnValue;value",
+        "My.Qltest.C;Generic<>;false;StepGeneric2;(S);;Argument[0];ReturnValue;value"
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/dataflow/external-models/steps.ql:24]: I just noticed this: `My.Qltest.C` is actually not a namespace, but a container class. Is this expected? Shouldn't we instead have `My.Qltest` as a namespace, and `C+Generic<>` as the class name?

Comment (tamasvajk) [reply csharp/ql/test/library-tests/dataflow/external-models/steps.ql:24]: Java seems to be using this notation for nested classes: `Base64$Decoder`. See https://github.com/github/codeql/blob/main/java/ql/src/semmle/code/java/dataflow/ExternalFlow.qll#L256-L263.

Comment (hvitved) [reply csharp/ql/test/library-tests/dataflow/external-models/steps.ql:24]: You are right. I will change this.

File: csharp/ql/test/library-tests/unification/Unification.expected [hunk 2]
```diff
@@ -44,19 +44,19 @@ constrainedTypeParameterSubsumes
...
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:38:11:38:22 | Nested<System.Int32>+NestedA<T11> |
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:38:11:38:22 | Nested<System.String>+NestedA<Int32> |
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:38:11:38:22 | Nested<System.String>+NestedA<T11> |
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:39:11:39:17 | Nested<>+NestedB |
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:39:11:39:17 | Nested<System.Int32>+NestedB |
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:39:11:39:17 | Nested<System.String>+NestedB |
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:41:22:41:33 | Nested<>+NestedB+NestedC<T12> |
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:41:22:41:33 | Nested<System.Int32>+NestedB+NestedC<Boolean> |
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:41:22:41:33 | Nested<System.Int32>+NestedB+NestedC<T12> |
+| Unification.cs:9:10:9:11 | T3 | Unification.cs:41:22:41:33 | Nested<System.String>+NestedB+NestedC<Decimal> |
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/unification/Unification.expected:58]: Not all type names show up here with fully qualified names. It looks like nested constructed generics are not handled properly.

File: csharp/ql/test/library-tests/overrides/Overrides22.expected [hunk 3]
```diff
@@ -36,15 +36,15 @@
...
 | overrides.D.f2() | overrides.A.f2() | overrides |
 | overrides.E2.M() | overrides.E.M() | overrides |
 | overrides.E.M() | overrides.I.M() | implements |
 | overrides.F.M() | overrides.I.M() | implements |
 | overrides.G2.M<S>(string, S) | overrides.G.M<S>(string, S) | overrides |
-| overrides.G.M<S>(string, S) | overrides.I2<String>.M<S>(string, S) | implements |
-| overrides.H<>.M<S>(TA, S) | overrides.I2<TA>.M<S>(TA, S) | implements |
-| overrides.Outer<>.A10.M<T>(Inner) | overrides.Outer<>.I6.M<T>(Inner) | implements |
+| overrides.G.M<S>(string, S) | overrides.I2<System.String>.M<S>(string, S) | implements |
+| overrides.H<>.M<S>(TA, S) | overrides.I2<overrides.H<>.<0>>.M<S>(TA, S) | implements |
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/overrides/Overrides22.expected:49]: I'm surprised to see `TA` being changed to `overrides.H<>.<0>`, but `S` not being changed. Also, shouldn't all instances of `TA` be replaced by `overrides.H<>.<0>`?

**Comment by tamasvajk** [comment]
Wouldn't it make sense to store the fully qualified names in the DB (if there's an easy built in way of getting it from Roslyn)? 

I did a quick search, and couldn't find any APIs for getting it, but [`SymbolDisplayFormat`](https://sourceroslyn.io/#Microsoft.CodeAnalysis/SymbolDisplay/SymbolDisplayFormat.cs,4e73ca1360ce043f) seems promising, and at the same time it already [shows](https://sourceroslyn.io/#Microsoft.CodeAnalysis/SymbolDisplay/SymbolDisplayGenericsOptions.cs,22) that even Roslyn uses `<>` notation when displaying "qualified names" of constructed generics types. So the current QL implementation  is quite close to theirs, the only difference seems to be the type parameters. [Here](https://sourceroslyn.io/#Microsoft.CodeAnalysis/SymbolDisplay/SymbolDisplayFormat.cs,133)'s an instance of `SymbolDisplayFormat` that we could use if we decided to move this logic to the extractor.

Here are some examples of type names in Roslyn tests:
* [With generics and nested types](https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Emit.UnitTests/CodeGen/CodeGenExplicitImplementationTests.cs,676)
* [With type parameters](https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Emit.UnitTests/CodeGen/CodeGenExplicitImplementationTests.cs,615)
* [With aliases, which we probably don't handle currently](https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Emit.UnitTests/CodeGen/CodeGenExplicitImplementationTests.cs,1064)

File: csharp/ql/src/semmle/code/csharp/Type.qll [hunk 4]
```diff
@@ -55,17 +55,50 @@ private predicate isObjectClass(Class c) { c instanceof ObjectType }
...
+      )
+  }
+
   /**
    * Holds if this type has the qualified name `qualifier`.`name`.
    *
    * For example the class `System.IO.IOException` has
    * `qualifier`=`System.IO` and `name`=`IOException`.
    */
   override predicate hasQualifiedName(string qualifier, string name) {
```
[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/ql/src/semmle/code/csharp/Type.qll:86]: We discussed this in person. We're sticking to the current implementation (such as `C<System.Int32>`) to keep the qualified names closer to the familiar C# syntax.

**Comment by tamasvajk** [comment]
> Wouldn't it make sense to store the fully qualified names in the DB (if there's an easy built in way of getting it from Roslyn)?
> 
> I did a quick search, and couldn't find any APIs for getting it, but [`SymbolDisplayFormat`](https://sourceroslyn.io/#Microsoft.CodeAnalysis/SymbolDisplay/SymbolDisplayFormat.cs,4e73ca1360ce043f) ...

We discussed this in person: Our conclusion was to not use the fully qualified name returned at runtime, and also not match exactly how Roslyn represent qualified names. We're actually quite close to the one implemented by Roslyn, here are some samples of the representation used by us:
* nested types: `NS1.C1+C2`
* constructed generics: `NS1.C1<System.Int32,System.Int32>`
* unbound generics: `NS1.C1<,>`
* open generics: `NS1.C1<T1,T2>`.

**Review by tamasvajk — APPROVED** [review]
LGTM

**Pull request merged**